util/queue.go

```diff
@@ -0,0 +1,58 @@
+package util
+
+import (
+	"errors"
+	"sync"
+)
+
+type Queue struct {
+	Items []interface{}
+	mu    sync.RWMutex
+}
+
+func NewQueue() Queue {
+	return Queue{
+		Items: []interface{}{},
+	}
+}
+
+func (q *Queue) Enqueue(item interface{}) error {
+	if item == nil {
+		return errors.New("cannot enqueue nil item")
+	}
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	q.Items = append(q.Items, item)
+
+	return nil
+}
+
+func (q *Queue) Dequeue() interface{} {
+	if len(q.Items) == 0 {
+		return nil
+	}
```

Comment:
this should be within a lock, or you could call isEmpty.

Reply:
done